pymeasure/adapters/adapter.py

```diff
@@ -41,37 +43,75 @@ class Adapter:
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
         if self.connection is not None:
             self.connection.close()
 
+    # Directly called methods.
     def write(self, command):
-        """ Writes a command to the instrument
+        """
+        Write a command to the instrument.
 
         :param command: SCPI command string to be sent to the instrument
         """
-        raise NameError("Adapter (sub)class has not implemented writing")
+        self.log.debug(f"WRITE:{command.encode()}")
```

Comment:
why do you `encode` here? At this point a string is passing by, does not `encode()` potentially "muddy the water"?

Reply:
The logging is tied together with the Test Generator.
Written/Read bytes have to be parsed.
I have two options:

1 mark the two types differently: "WRITE:a funny string" and "WRITEB:b'\x00some bytes\x03jh67'"
2 or write always bytes, as in the end, strings will be sent as bytes to the device anyway.

I was not yet fixed on one or the other approach, but you hinted advantageds of the second one

Reply:
Logging is done with strings, I think. Once the bytes are expressed as strings, they have to be parsed, as there is no other method (I could find).

Optionally the adapter could open byte streams, such that we do not have to decode strings to bytes anymore.

Reply:
At least [one person](https://stackoverflow.com/a/67421044/599884) has written a custom handler to log as bytes. I have not investigated if that could fit our situation, but the use case seems at least adjacent (feeding log output to `serial`)

Reply:
Thanks for that idea of a bytes handler. With a byte stream, we keep bytes objects and do not have to parse the string.